### ISSUE TITLE
AuthSession returns dismiss result even before the browser is opened

### ISSUE THREAD

**Issue by ivansenic** [expo/expo#6679]
## 🐛 Bug Report

Calling the `AuthSession.startAsync` for the first time returns the `dismiss` result immediately after calling. The browser opens correctly in the meantime, but any action done afterwards in the browser is ignored (login correctly, cancel, etc). This means that our users will go to the login screen, login successfully, but the app would understand this session as being dismissed.

In addition, but not connected to this issue directly:
* On Android, when the user cancels the login (does not open browser, clicks on `X` in browser, etc) the result sent back to us is also `dismiss` as well. This means we can not distinguish between cancel and dismiss, which prevents us to create at least a meaningful workaround for this issue.
* Calling  `AuthSession.dismiss` before starting the session crashes the app (in the Expo client)

### Environment

**IMPORTANT**:
* only happens on Android
* only happens in production mode or the built standalone app **(not reproducible in the dev mode)**
* only happens on the first call to the `AuthSession.startAsync`, successive calls bring the expected behaviour
* confirmed in both simulator and Android device

```
  Expo CLI 3.11.3 environment info:
    System:
      OS: Linux 5.0 Ubuntu 18.04.3 LTS (Bionic Beaver)
      Shell: 4.4.20 - /bin/bash
    Binaries:
      Node: 10.18.0 - /usr/bin/node
      npm: 6.13.4 - /usr/bin/npm
    npmPackages:
      @types/react: ~16.9.0 => 16.9.16 
      @types/react-native: ~0.60.23 => 0.60.25 
      @types/react-navigation: ^3.0.7 => 3.0.7 
      expo: ^36.0.2 => 36.0.2 
      react: ~16.9.0 => 16.9.0 
      react-native: https://github.com/expo/react-native/archive/sdk-36.0.0.tar.gz => 0.61.4 
      react-navigation: ^3.13.0 => 3.13.0 
    npmGlobalPackages:
      expo-cli: 3.11.3
```

### Steps to Reproduce

Simply call `AuthSession.startAsync` for the first time since the app has been loaded.

### Expected Behavior

I would expect that dismissed is not the result, especially that the browser get open and user not knowing what s going proceeds to the login page.

### Reproducible Demo
Here is the most simple demo: https://snack.expo.io/rJZ!LryxI

However this needs to be run in the production mode in order to actually get the bug. I am not sure if expo snack supports production mode.

### Extra

* I am also linking here a question from StackOverflow from another user having the same issue: https://stackoverflow.com/questions/59589158/expo-authsession-immediately-resolves-as-dismissed
* I guess this can be connected to https://github.com/expo/expo/issues/6289

**Comment by LucaColonnello** [comment]
I've been debugging this and I think I tracked down the issue here.

**The problem**
In Android and iOS < 11, expo-web-browser used by AuthSession, is using a polyfill version of the NativeAuthSession behaviour available in iOS >= 11.

_The way this works is by creating a race condition waiting for one of the following to happen first:_

1. user comes back from the just opened browser to the app, trigger an AppState change (in this case user is assumed to have clicked dismiss or have closed the browser or have come back to the app)
2. redirect to the returnUrl (default expo app url as intent) as a callback of a successful login action in the browser window (in this case we assume login was successful)

When everything is ok, 2 should happen before 1 and that will dismiss altogether the race condition, resulting in `{ type: 'success' }`.

**What happens here, you can notice debugging, is that the first promise in the race condition throws before opening the browser. The reason is that AppState change triggers a `change` event with state `active` before the browser activity is opened, which means that the change event is triggered without 
a real AppState change.**

This behaviour is actually by design when no other listener to LifecycleEvent has ever been created, as the initial state will be null before AppState grabs it from the bridge (you can verify this by adding another usage of AppState in your app before calling AuthSession features).

The first listener to a LifecycleEvent gets called as soon as it's added even if state hasn't changed, to notify of the initial value.


Here's a Snack to demonstrate what I'm saying https://snack.expo.io/H1W9Lllg8...
By calling AppState in the render here, I'm triggering AppState to listen for changes earlier, without suffering from the first call handler call (although as said above is not something we can see in Expo Snack, but I used this to solve the same issue in my app and it works consistently everywhere)...

**References**
- WebBrowser race condition https://github.com/expo/expo/blob/master/packages/expo-web-browser/src/WebBrowser.ts#L179-L182
- AppState listener returning dismiss when user comes back from browser activity https://github.com/expo/expo/blob/master/packages/expo-web-browser/src/WebBrowser.ts#L138-L161
- ReactContext LifecycleEventListener interface explaining the design of AppState event emitters https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/bridge/LifecycleEventListener.java#L33-L35 ; https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/bridge/ReactContext.java#L165-L192

**Solution**
I have no idea whether this is something that should be taken with ReactNative to argue the design choice of AppState, but the solution for Expo could be to change this https://github.com/expo/expo/blob/master/packages/expo-web-browser/src/WebBrowser.ts#L134-L142 in the following way:

```js
// Store the `resolve` function from a Promise to fire when the AppState
// returns to active
let _onWebBrowserCloseAndroid: null | (() => void) = null;

// Store previous app state to check whether the listener has ever been attached
let _previousAppState: null | string = AppState.currentState;

function _onAppStateChangeAndroid(state: AppStateStatus) {
  // if _previousAppState is null, we assume that the first call to
  // AppState#change event is not actually triggered by a real change
  // (https://facebook.github.io/react-native/docs/appstate#basic-usage)
  if (_previousAppState === null) {
    _previousAppState = state;
    return;
  }

  if (state === 'active' && _onWebBrowserCloseAndroid) {
    _onWebBrowserCloseAndroid();
  }
}
```

Please let me know if this is ok, I could make a PR or we could discuss further...

**Comment by schellack** [comment]
This same issue has been plaguing me for months now, but I had not yet been able to get to root cause. I used the patch to Expo WebBrowser that @LucaColonnello posted above, tested it in my app, and it seems to work perfectly.

**Comment by LucaColonnello** [comment]
@schellack glad to hear it helped you 😊

Quick question: did you try the specific patch to the event listener with the previous state var in Expo WebBrowser or the hack of putting AppState in your app first (I think you said you tried the patch but just to be 100% clear)?

Thanks!

**Comment by schellack** [comment]
@LucaColonnello I patched Expo WebBrowser. I already have another framework for authentication that uses expo-web-browser's AuthSession and hooks into AppState to do so. I didn't really want to have a third AppState event listener to manage, especially just as a hack. 

I've also seen this issue occur intermittently. Most of the time the issue does occur (like 9 times out of 10), but every now and then it seems that the race condition allows redirect to come back first and things work as they should. The past inconsistency has made me not want to add another variable to the mix. Plus the actual patch to WebBrowser was simple 😃

**Comment by LucaColonnello** [comment]
@schellack so if you confirm the patch I suggested works, I’m going to open a PR contributing here 😁

**Comment by ivansenic** [comment]
My kudos to @LucaColonnello, I rarely stabmle to such a quick and effective reaction on the bug ticket.. Workaround provided, fix provided, what can you expect more :clap: P.S. If you ever come to Belgrade I am taking you out for a drink, just ping me :beers:

**Comment by schellack** [comment]
So I downgraded to Expo SDK 35 and expo-web-browser 7.0.1 in an effort to see if the upgrade to Expo SDK 36 was related to another issue (#6693), which has no relation to this one. After doing so, it just so happens that the patch doesn't seem to work any more. 

As far as I can tell, nothing changed in expo-web-browser between 7.0.1 and 8.0.0 related to this, and the patch is applied the same way. So I'm doing some more digging to see why.

**Comment by LucaColonnello** [comment]
@schellack interesting, thanks for sharing! Might be worth checking ReactNative version in Expo SDK 35...

Keep us updated!

**Comment by DianaKoenraadt** [comment]
Great to see a pull request has been made, thanks! Where will this appear once the PR is unblocked, will it appear on SDK 36 or in the npm package `expo-web-browser`?

**Comment by robingrella** [comment]
Is this fix also solving the WebBrowser.openAuthSessionAsync() function for android and ios <=11? I still get type: dismiss as the result of openAuthSessionAsync(), even if the authentication was succesful.

**Comment by LucaColonnello** [comment]
@robingrella 
> I still get type: dismiss

Is this after you tried my PR?

**Comment by robingrella** [comment]
yes, I tried it with your changes.

**Comment by actuallymentor** [comment]
Is there any progress on this issue? I just lost half a day diagnosing what was going on and ending up here...

**Comment by LucaColonnello** [comment]
> Is there any progress on this issue? I just lost half a day diagnosing what was going on and ending up here...

Unfortunately I have not been able to move forward. I had several issues with the CI. I’ve been recommended a way forward which I need to try, but this is also missing test which are a bit complicated for this use case.

I’ll try again and see what I can get out of this but I need support as I’m new to the codebase and general testing in the react-native area...

**Comment by LucaColonnello** [comment]
@actuallymentor could you in the meanwhile try the fix suggested in your app?

It works in our app but some people said it didn’t for them...

Many thanks!

**Comment by LucaColonnello** [comment]
@robingrella I pushed new changes and made a build of the fix.
Do you want to try now?

Missing build might have been culprit for the fix not to work properly when you linked the branch...

**Comment by actuallymentor** [comment]
> could you in the meanwhile try the fix suggested in your app?

I ended up implementing a "something broke, not your fault, please try again" message. As a rule I don't edit underlying packages to prevent legacy issues.

> I pushed new changes and made a build of the fix. Do you want to try now?

Which package did you publish to? I'm not seeing updates to any packages that seem relevant to this.

**Comment by eikonomega** [comment]
@LucaColonnello Thanks so much for your work on this. I have a question. Do you need to do anything other than update the contents of `node_modules/expo-web-browser/src/WebBrowser.ts` and republish the app? I still seem to have the double login issue on Android after doing this.

**Comment by LucaColonnello** [comment]
@eikonomega I think you might need to make your changes in the build folder as opposed to src folder...

**Comment by simplesthing** [comment]
I am having this same issue with a fresh install of all expo packages, when I inspect code I see the old code using - "name": "expo-web-browser",  "version": "8.2.1", was there a regression or the code was not merged into the correct place?

**Comment by JuanDavidLopez95** [comment]
Hi @simplesthing ,

I was encountering the same issue as you on Android. 

Using `AuthSession.startAsync()`, the result of this function always returned `{ type: 'dismiss" } `even after successful authentication. I was able to get this fixed by supplying a `returnUrl` to the config object parameter passed to `AuthSession.startAsync()`. The `AuthSession.startAsync()` object parameters are outlined [here](        https://docs.expo.io/versions/latest/sdk/auth-session/#authsessionstartasyncoptions
).

It is strange but explicity supplying a `returnUrl` parameter, perhaps with a value of `AuthSession.makeRedirectUri()`, did the trick for me.

**Comment by simplesthing** [comment]
@JuanDavidLopez95  Thank you for your response, i have read that issue as well, I am confused as to what the `./helpers` file links to and what the `getAuthorizeUrl` is doing, I have serarched for this in documentation I do not find anything. I am supplying a redirect uri  and everything logs correctly.  When I Google `makeRedirectUri` all I find is an issue that this is not included on expo. https://github.com/expo/expo/issues/8073

` const authUrl = state.settings.loginServiceUrl
    const redirectUri = Linking.makeUrl('auth/in')
    dispatch({type: 'log', payload: 'onLogin ' + AppState.currentState + ' ' + redirectUri})
    try {
      const result = await startAsync({
        authUrl,
        redirectUri
      })
      dispatch({type: 'log', payload: `result: ${JSON.stringify(result)}`})
    } catch(e) {
      dispatch({type: 'log', payload: `error ${e}`})
    }`

**Comment by YoranBrondsema** [comment]
I'm having an issue with `AuthSession` as well, but with `useAuthRequest`. I described it on the [Expo forum](https://forums.expo.io/t/useauthrequest-always-returns-dismiss/38476) but mentioning it here in case anyone has any ideas. I tried the `AppState.currentState` "trick" but that didn't work for me.

**Comment by viantirreau** [comment]
Hi @YoranBrondsema!
Did you figure it out? I'm having the same exact problem when implementing 'sign-up with GitHub' via OAuth.

**Comment by YoranBrondsema** [comment]
@viantirreau I haven't :/

**Comment by viantirreau** [comment]
@YoranBrondsema 
I finally solved it using `AuthSession.startAsync`! I had to adapt [this older Snack](https://snack.expo.io/@bacon/github?session_id=snack-session-uJ5o9JtYE&referrer=https%3A%2F%2Fblog.expo.io%2Ffirebase-github-authentication-with-react-native-2543e32697b4), but in the end it just worked fine for my purposes (OAuth). Just make sure to setup the `REDIRECT_URL` to something on the lines of `https://auth.expo.io/@<username>/<app-name>`. Although it's more involved, as you have to manually configure some of the urls, I think you can make it work on itsme's OIDC using this method.

Good luck!

**Comment by YoranBrondsema** [comment]
@viantirreau Great, thanks for sharing your solution. I'm not working on this part of the application at the moment but will remind myself to let it know here whether the solution works for me.

**Comment by tylerthedeveloper** [comment]
> @YoranBrondsema
> I finally solved it using `AuthSession.startAsync`! I had to adapt [this older Snack](https://snack.expo.io/@bacon/github?session_id=snack-session-uJ5o9JtYE&referrer=https%3A%2F%2Fblog.expo.io%2Ffirebase-github-authentication-with-react-native-2543e32697b4), but in the end it just worked fine for my purposes (OAuth). Just make sure to setup the `REDIRECT_URL` to something on the lines of `https://auth.expo.io/@<username>/<app-name>`. Although it's more involved, as you have to manually configure some of the urls, I think you can make it work on itsme's OIDC using this method.
> 
> Good luck!


what is the format of authUrl for AuthSession.startAsync?

**Comment by rikkitissier** [comment]
I'm experiencing this same problem when using `WebBrowser. openAuthSessionAsync`. Everything works in dev mode and in the production iOS app, but fails in the production Android app. Through logging I can see that `{ type: 'dismiss' }` is returned before the browser even shows, just as described here. The suggested workaround of logging `AppState.currentState` unfortunately seems to no longer work. I'm really at a loss as to where to go from here - I can't release our app without authentication on Android :/

**Comment by erikkjernlie** [comment]
Did you solve this @rikkitissier?

**Comment by actuallymentor** [comment]
Not sure why this closed, I still experience the issue in production.

**Comment by StefanoZaghi1987** [comment]
Hello, do you have any updates about this issue?
I'm still experiencing it on production Android environment.

I found it installing an Expo app as an Android standalone app.
Everything is fine, instead, by running the app through Expo.

As explained before by @LucaColonnello, when the app is running on Android as a standalone app, every authentication attempt returns `dismiss`.

I'm having this issue both using `AuthSession.useAuthRequest` ([https://docs.expo.io/versions/latest/sdk/auth-session/#useauthrequest](url)) and also using `AuthSession.startAsync` ([https://docs.expo.io/versions/latest/sdk/auth-session/#authsessionstartasyncoptions](url)).

I am currently using the following package versions:
- `"expo": "~40.0.1"`
- `"react": "16.13.1"`
- `"react-native": "https://github.com/expo/react-native/archive/sdk-40.0.1.tar.gz"` `(0.63.4)`
- `"expo-auth-session": "~3.0.0"`
- `"expo-web-browser": "~8.6.0"`
- `"expo-linking": "~2.0.1"`.

I tried both to log `AppState.currentState` and the patch suggested by @LucaColonnello, but everything was no use.

Please, can you help me?

I'll try to explain here some details...

My app authentication flow is based on Microsoft Azure Active Directory Authentication ([https://docs.expo.io/guides/authentication/#azure](url)).

We can start with some details about my Azure AD configuration:

![image](https://user-images.githubusercontent.com/30727653/108521458-094f0300-72cc-11eb-84fd-26351316361f.png)

![image](https://user-images.githubusercontent.com/30727653/108521705-4c10db00-72cc-11eb-844e-c6821e49c0b2.png)

Here is how i tried to authenticate using `AuthSession.useAuthRequest` and `AuthSession.startAsync`.

### `AuthSession.useAuthRequest`

In this paragraph I'll show how I tried to authenticate with `useAuthRequest` hook and `promptAsync`.
Everything works on iOS and on Android dev environment.
I'm having the `type='dismiss'` issue, instead, on Android production environment.

```jsx
import * as React from 'react';
import { useEffect } from 'react';
import * as WebBrowser from 'expo-web-browser';
import { useAutoDiscovery, useAuthRequest } from 'expo-auth-session';

import { authConfig, AzureEndpointUrl } from '../constants/Authentication';
import { handleAuthenticationResponse } from '../core/authentication/Authentication.base';

import { Text, TouchableOpacity, View, SafeAreaView } from '../components/Themed';
import { styles } from '../layout/Styles';

WebBrowser.maybeCompleteAuthSession();

export default function ExpoSignInScreen({navigation, onResponseReceived, onSignIn}) {
  // Authentication request config
  const requestConfig = {
    clientId: authConfig.CLIENT_ID,
    scopes: authConfig.SCOPES,
    redirectUri: authConfig.REDIRECT_URI
  };

  // OpenID discovery document
  const authDiscoveryDocument = useAutoDiscovery(AzureEndpointUrl);

  // Authentication Request
  const [request, response, promptAsync] = useAuthRequest(
    requestConfig,
    authDiscoveryDocument
  );

  // Authentication response management
  useEffect(() => {
    if (response) {
      console.log(response);
      handleAuthenticationResponse(response, onResponseReceived, onSignIn);
    }
  }, [response]);

  return (
    <SafeAreaView style={[styles.verticalContainer, styles.contentContainer]}>
      ...
      <View style={styles.verticalContainer}>
        <TouchableOpacity style={styles.button} disabled={!request} onPress={() => { promptAsync({ useProxy: false }); }}>
            <Text style={styles.buttonText}>Sign In</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.button} disabled={!request} onPress={() => { promptAsync({ useProxy: true }); }}>
            <Text style={styles.buttonText}>Sign In (use proxy)</Text>
        </TouchableOpacity>
        ...
      </View>
    </SafeAreaView>
  );
}
```

### `AuthSession.startAsync`

In this sectionI'll show how I tried to authenticate with `AuthSession.startAsync`.
I tried both to leave the `returnUrl` as default and also to pass it.

So, using the utilities

```jsx
import { startAsync } from 'expo-auth-session';

export async function startAuthRequestAsync(requestUrl) {
    return await startAsync({ authUrl: requestUrl });
};

export async function startAuthRequestAndReturnAsync(requestUrl, returnUrl) {
    return await startAsync({ authUrl: requestUrl, returnUrl: returnUrl });
};
```

and

```jsx
// Sign-in and get and ID token
export function getIdTokenUrl(authorizationEndpoint, client_id, redirect_uri, scope, state, nonce) {
    return authorizationEndpoint + `?client_id=${client_id}&response_type=id_token&redirect_uri=${encodeURIComponent(redirect_uri)}&response_mode=fragment&scope=${encodeURIComponent(scope)}&state=${state}&nonce=${nonce}`;
}

// Sign-in and get both an ID token and an authorization code
export function getIdTokenAndAuthorizationCodeUrl(authorizationEndpoint, client_id, redirect_uri, scope, state, nonce) {
    return authorizationEndpoint + `?client_id=${client_id}&response_type=${encodeURIComponent('id_token code')}&redirect_uri=${encodeURIComponent(redirect_uri)}&response_mode=fragment&scope=${encodeURIComponent(scope)}&state=${state}&nonce=${nonce}`;
}

// Sign-in and get both an ID token and an access token
export function getIdTokenAndAccessTokenUrl(authorizationEndpoint, client_id, redirect_uri, scope, state, nonce) {
    return authorizationEndpoint + `?client_id=${client_id}&response_type=${encodeURIComponent('id_token token')}&redirect_uri=${encodeURIComponent(redirect_uri)}&response_mode=fragment&scope=${encodeURIComponent(scope)}&state=${state}&nonce=${nonce}`;
}

// Sign Out
export function getSignOutUrl(endSessionEndpoint, post_logout_redirect_uri) {
    return endSessionEndpoint + `?post_logout_redirect_uri=${encodeURIComponent(post_logout_redirect_uri)}`;
}
```

we have

```jsx
import * as React from 'react';
import { AppState } from 'react-native';
import Constants from 'expo-constants';
import * as WebBrowser from 'expo-web-browser';
import { useAutoDiscovery } from 'expo-auth-session';

import { authConfig, AzureEndpointUrl } from '../constants/Authentication';
import { getScopes, handleAuthenticationResponse } from '../core/authentication/Authentication.base';
import { startAuthRequestAsync, startAuthRequestAndReturnAsync } from '../core/authentication/Authentication.Expo';
import { getIdTokenAndAccessTokenUrl } from '../core/authentication/Authentication.Implicit';

import { Text, TouchableOpacity, View, SafeAreaView } from '../components/Themed';
import { styles } from '../layout/Styles';

WebBrowser.maybeCompleteAuthSession();

export default function OpenIDSignInScreen({navigation, onResponseReceived, onSignIn}) {
  // Authentication request config
  const requestConfig = {
    clientId: authConfig.CLIENT_ID,
    scopes: authConfig.SCOPES,
    redirectUri: authConfig.REDIRECT_URI
  };

  // OpenID discovery document
  const authDiscoveryDocument = useAutoDiscovery(AzureEndpointUrl);

  let response = null;

  const getIdTokenAsync = async () => {
    const requestUrl = getIdTokenAndAccessTokenUrl(authDiscoveryDocument.authorizationEndpoint, authConfig.CLIENT_ID, authConfig.REDIRECT_URI, getScopes(authConfig.SCOPES), 'Aa123456', 'Aa78910');
    response = await startAuthRequestAsync(requestUrl);

    if (response) {
      console.log(response);
      handleAuthenticationResponse(response, onResponseReceived, onSignIn);
    }
  };

  const getIdTokenAndDefaultReturnAsync = async () => {
    const requestUrl = getIdTokenAndAccessTokenUrl(authDiscoveryDocument.authorizationEndpoint, authConfig.CLIENT_ID, authConfig.REDIRECT_URI, getScopes(authConfig.SCOPES), 'Aa123456', 'Aa78910');
    response = await startAuthRequestAndReturnAsync(requestUrl, `${Constants.linkingUrl}expo-auth-session`);

    if (response) {
      console.log(response);
      handleAuthenticationResponse(response, onResponseReceived, onSignIn);
    }
  };

  const getIdTokenAndReturnAsync = async () => {
    const requestUrl = getIdTokenAndAccessTokenUrl(authDiscoveryDocument.authorizationEndpoint, authConfig.CLIENT_ID, authConfig.REDIRECT_URI, getScopes(authConfig.SCOPES), 'Aa123456', 'Aa78910');
    response = await startAuthRequestAndReturnAsync(requestUrl, authConfig.REDIRECT_URI);

    if (response) {
      console.log(response);
      handleAuthenticationResponse(response, onResponseReceived, onSignIn);
    }
  };

  return (
    <SafeAreaView style={[styles.verticalContainer, styles.contentContainer]}>
      ...
      <View style={styles.verticalContainer}>
        <TouchableOpacity style={styles.button} onPress={getIdTokenAsync}>
            <Text style={styles.buttonText}>Sign In</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.button} onPress={getIdTokenAndDefaultReturnAsync}>
            <Text style={styles.buttonText}>Sign In (with default return)</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.button} onPress={getIdTokenAndReturnAsync}>
            <Text style={styles.buttonText}>Sign In (with return)</Text>
        </TouchableOpacity>
        ...
      </View>
    </SafeAreaView>
  );
}
```

Everything works on iOS and on Android dev environment.
I'm having the `type='dismiss'` issue, instead, on Android production environment.

Finally, I tried again starting the authentication session with `WebBrowser.openAuthSessionAsync`, but it was useless.
The result was the same as before.

Please, can you help me to overcome this issue?

**Comment by eprice122** [comment]
Issue still present

**Comment by tajetaje** [comment]
@brentvatne @ivansenic Even if there is no solution to this could we at least get this issue reopened as it is linked to in a few other issues? Over half the discussion has been after this issue was closed.

**Comment by rossparachute** [comment]
Hey folks,

Still an issue in April 2022. 

### Environment
This issue occurs on Android standalone and on published Android releases on Expo Go. Both emulator and real devices.

A locally ran build via the Metro bundler does not appear to have the issue. Making it more difficult to debug.

`Expo CLI 5.0.2 environment info:
    System:
      OS: Windows 10 10.0.19044
    Binaries:
      Node: 16.13.0 - C:\Program Files\nodejs\node.EXE
      Yarn: 1.22.5 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
      npm: 8.1.0 - C:\Program Files\nodejs\npm.CMD
    IDEs:
      Android Studio: Version     2020.3.0.0 AI-203.7717.56.2031.7784292
    npmPackages:
      expo: ^44.0.0 => 44.0.6
      react: 17.0.1 => 17.0.1
      react-dom: 17.0.1 => 17.0.1
      react-native: 0.64.3 => 0.64.3
      react-native-web: 0.17.1 => 0.17.1
    Expo Workflow: managed`

`"expo-auth-session": "~3.5.0",
"expo-random": "~12.1.1",`

### Steps to Reproduce
1. Cold boot app so there's nothing in memory regarding a successful authentication
2. Destructure the expo-authsession Google.useAuthRequest() hook to get your promptAsync()
3. Call that promptAsync() method when you want to authenticate, e.g - on button press
4. Successfully authenticate with your Google account
5. You will notice the browser won't automatically close on success and when you close it yourself, it is treated as a dismiss and therefore the process fails

### Expected Behavior
We would expect that once a Google account is successfully authenticated the browser window would close automatically with the response object's type property from Google.useAuthRequest() being marked as "success" along with the resulting token.

### Actual Behaviour
Even when the Google account is authenticated in the browser, the window will not close automatically. This is treated as a "dismiss" type in the response when the user closes it themselves.

### Example Code (shortened for brevity)
`
import React, { useState, useEffect } from "react";
import { Button } from "react-native";

// Authentication
import * as WebBrowser from 'expo-web-browser';
import * as AuthSession from 'expo-auth-session';
import * as Google from 'expo-auth-session/providers/google';

WebBrowser.maybeCompleteAuthSession();

export const SocialLoginForm = ({}) => {
  // Authentication setup
  const [googleReq, googleRes, googlePromptAsync] = Google.useAuthRequest({
    ...GOOGLE_AUTH_SETTINGS,
  });

  // Handlers
  const onGoogleSubmit = async () => {
    googlePromptAsync();
  } 

  // Hooks
  useEffect(() => {
    if (googleRes?.type === 'success') {
      // authenticate using the googleRes.authentication object
      return;
    }

    if(googleRes?.type === 'dismiss') {
      // show message
      return;
    }

    if(googleRes && (googleRes.errorCode || googleRes.error))
      // handle error message
  }, [googleRes]);

  return (
    <Button
        text="Google Login"
        onPress={() => onGoogleSubmit()}
    />
  )
}
`

### Extra
We've heard that mileage may vary by trying the deprecated authentication flow. We'll give that a go and report back.

Discussed here:
https://www.youtube.com/watch?v=hmZm_jPvWWM

Deprecated flow:
https://www.youtube.com/watch?v=QT0PXhH9uTg

**Comment by travishoki** [comment]
Still experiencing this issue.

**Comment by tajetaje** [comment]
> Still experiencing this issue.

Are you on SDK 45? I was hoping that the updates to WebBrowser would solve this

**Comment by rossparachute** [comment]
@tajetaje Will give that a go and let you know!

**Comment by travishoki** [comment]
> 
I'm using Expo SDK: 43.0.0. I could try upgrading. I was having issues with audio last time I tried upgrading so I've been a hesitant. I'll give it a go and report back!

**Comment by travishoki** [comment]
> 

@tajetaje I upgraded to Expo SDK 45 and am still having issues with the standalone Android app. :(

This is actually my issue:
WebBrowser.openAuthSessionAsync returns "dismiss" even when the user successfully logged in #6289 https://github.com/expo/expo/issues/6289

On the standalone Android version, the login seems to work, but when coming back to the app it isn't resolving correctly. It seems as though the login was rejected.

**Comment by tajetaje** [comment]
> > 
> 
> 
> 
> @tajetaje I upgraded to Expo SDK 45 and am still having issues with the standalone Android app. :(
> 
> 
> 
> This is actually my issue:
> 
> WebBrowser.openAuthSessionAsync returns "dismiss" even when the user successfully logged in #6289 
> 
> 
> 
> On the Android standalone app, the login seems to work, but when coming back to the app it is returning with {type: “dismiss”}. It works perfectly on iOS and in the emulator but not on the Android standalone app.

Yeah I suspect what I noticed in #16712 is still going on

**Comment by mraible** [comment]
I've been struggling with this for several hours. [This solution](https://github.com/expo/examples/issues/195#issuecomment-655212634) worked for me.

**Comment by pop-thanit** [comment]
> ```js
> ResponseRe
> ```

Does it work with Google-auth ? 
I followed [this](https://github.com/chelseafarley/google-auth-demo/blob/master/App.js) step

This works fine on iOS standalone app but on Android every authentication attempt returns dismiss.

Can you share the example code ?

**Comment by pop-thanit** [comment]
> I've been struggling with this for several hours. [This solution](https://github.com/expo/examples/issues/195#issuecomment-655212634) worked for me.

Does it work with Google-auth ?
I followed [this](https://github.com/chelseafarley/google-auth-demo/blob/master/App.js) step

This works fine on iOS standalone app but on Android every authentication attempt returns dismiss.

Can you share the example code ?

**Comment by knorsen** [comment]
@travishoki did you ever find a solution? Running into exactly the same issue with Google auth on Android standalone. Would love to know if you solved it.

**Comment by travishoki** [comment]
> @travishoki did you ever find a solution? Running into exactly the same issue with Google auth on Android standalone. Would love to know if you solved it.

@knorsen Yes, I was able to get it working. I'm using Expo 45 and expo-auth-session. Go checkout the answer from @XiangZhang0216 on this thread https://github.com/expo/expo/issues/10860. His suggested configurations in app.json did the trick.

**Comment by knorsen** [comment]
Thanks so much @travishoki ! I can confirm the addition you linked to seems to be working for me as well.

**Comment by totallytotallyamazing** [comment]
@travishoki and @knorsen I've now tried your suggested remedy in sdk-46 and then sdk-45. I made sure to rebuild with "expo build:android" after editing my app.json file and also made the correct changes to my AndroidManifest.xml file. 
My dev environment is Mac Big Sir and so far only my "open web" server is working properly and loading the Google user data into my app, so that I'm not left holding the bag with the "Something went wrong trying to finish signing in. Please close ..." message, like I am in the "open Android" server.
Are there any suggestions that either of you guys can think of, or a specific version of React-Native I should use, etc. 
Thank you ahead of time!!

**Comment by knorsen** [comment]
@totallytotallyamazing unfortunately nothing much to suggest. I will say I did not make any edits to the manifest file like you mentioned above though. It sounds like you might be using a slightly different solution than the one we had success with. I definitely tried a lot of different approaches with no success before the one that finally did work, so if you are not following that example exactly I'd start there. It's a really frustrating issue.

**Comment by josafath597** [comment]
I tried what @XiangZhang0216 proposed but it is not working for me

![googleLoginError](https://user-images.githubusercontent.com/45771451/201432521-fb1c2ee4-93e0-4c1d-aa55-15efc68a7077.PNG)

However when I make the apk and log in with google the following happens

![1](https://user-images.githubusercontent.com/45771451/201433388-ae856859-84d9-4db6-ad81-46a824936b41.jpeg)

![WhatsApp Image 2022-11-11 at 3 25 27 PM](https://user-images.githubusercontent.com/45771451/201433406-d3d5ebd3-b751-4b75-a5de-ca2ed652fbdb.jpeg)

![3](https://user-images.githubusercontent.com/45771451/201433435-210a5d61-e6a7-49c8-ae2d-dcf2f75c7005.jpeg)

and it no longer returns to the app automatically it stays like this and when you close the browser by touching the X it returns to the app and marks "dismiss" again

I hope you can help me

**Comment by aggounix** [comment]
I get the same behavior as @josafath597 
Have you benn able to solve it please?

**Comment by totallytotallyamazing** [comment]
@aggounix and @josafath597 I did the xml edits that @XiangZhang0216 suggested, on this thread [https://github.com/expo/expo/issues/10860](https://github.com/expo/expo/issues/10860). Also make the following configuration changes to your app.json, 
(Note the "scheme' and "path" keys [Below], especially "scheme" were the secrete code part of the "ANDROID_CLIENT_ID" comes after the "com.googleusercontent.apps." part, this reverse order is important):
```JSON
,
    "android": {
      "adaptiveIcon": {
        "foregroundImage": "./assets/adaptive-icon.png",
        "backgroundColor": "#FFFFFF"
      },
      "package": "host.exp.exponent",
      "intentFilters": [
        {
          "action": "MAIN",
          "category": "LAUNCHER",
          "data": {
            "scheme": "com.googleusercontent.apps.ANDROID_CLIENT_ID",
            "path": "/oauth2redirect"
          }
        }
      ]
    },
```
Don't forget to add "redirectUri:" to your Google.useAuthRequest object (Below):
```JS
const [request, response, promptAsync] = Google.useAuthRequest({
        expoClientId: `${EXPO_CLIENT_ID}.apps.googleusercontent.com`,
        androidClientId: `${ANDROID_CLIENT_ID}.apps.googleusercontent.com`,
        iosClientId: `${IOS_CLIENT_ID}.apps.googleusercontent.com`,
        webClientId: `${WEB_CLIENT_ID}.apps.googleusercontent.com`,
        redirectUri: makeRedirectUri({
            scheme: 'host.exp.exponent',
            useProxy: true,
            native: Platform.select({
                android: `com.googleusercontent.apps.${ANDROID_CLIENT_ID}:/oauth2redirect`,
                default: undefined
            })
        }),
        scopes: ['profile', 'email'],
        permissions: ['public_profile', 'email', 'gender', 'location']
    });
```

Last but not least rebuild your EXPO build.

Good luck because this was tricky but I finally got it working after much trial and error and it just suddenly started working about 3 week after my original post on this chain (Oct. 2022) almost out of the blue, so I may have done something else too that I'm unaware of.

**Comment by Quingsley** [comment]
1.SDK Version: 47
2. Platforms(Android/iOS/web/all):
The Expo documentation clearly outlines how to setup Authentication with Google and AuthSession API, but only in development mode. The example given shows how to set up a project in Google Cloud Console (GCP) but does not give steps to setup the GCP in production mode.

I am facing a challenge when setting up GCP using an Expo managed application. When setting up GCP, you are required to provide Authorized JavaScript Origin and Authorized redirect URIs. I have given these as https://auth.expo.io and https://auth.expo.io/@{expo-username}/{slug}.

However, when going to production, GCP requires you to verify the domains you provided in the Google Search Console. In this case, the domain I am supposed to verify is expo.io, which I do not own.

Is there a way to solve  this issue?
![Screenshot 2023-04-27-105216](https://user-images.githubusercontent.com/74877189/234798067-7f505bea-3fd9-40df-b7bf-9691c57968fc.png)

![Screenshot 2023-04-27-105629](https://user-images.githubusercontent.com/74877189/234797973-8e2be67b-ef3c-45f7-8d98-867481b80e62.png)
